source/common/ext_authz/ext_authz_impl.cc

```diff
@@ -0,0 +1,218 @@
+                           const envoy::service::auth::v2::CheckRequest& request,
+                           Tracing::Span& parent_span) {
+  ASSERT(callbacks_ == nullptr);
+  callbacks_ = &callbacks;
+
+  request_ = async_client_->send(service_method_, request, *this, parent_span, timeout_);
+}
+
+void GrpcClientImpl::onSuccess(std::unique_ptr<envoy::service::auth::v2::CheckResponse>&& response,
+                               Tracing::Span& span) {
+  CheckStatus status = CheckStatus::OK;
+  ASSERT(response->status().code() != Grpc::Status::GrpcStatus::Unknown);
+  if (response->status().code() != Grpc::Status::GrpcStatus::Ok) {
+    status = CheckStatus::Denied;
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceUnauthz);
+  } else {
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceOk);
+  }
+
+  callbacks_->onComplete(status);
+  callbacks_ = nullptr;
+}
+
+void GrpcClientImpl::onFailure(Grpc::Status::GrpcStatus status, const std::string&,
+                               Tracing::Span&) {
+  ASSERT(status != Grpc::Status::GrpcStatus::Ok);
+  UNREFERENCED_PARAMETER(status);
+  callbacks_->onComplete(CheckStatus::Error);
+  callbacks_ = nullptr;
+}
+
+void ExtAuthzCheckRequestGenerator::addressToProtobufAddress(
+    envoy::api::v2::Address& proto_address, const Network::Address::Instance& address) {
+  if (address.type() == Network::Address::Type::Pipe) {
+    proto_address.mutable_pipe()->set_path(address.asString());
+  } else {
+    ASSERT(address.type() == Network::Address::Type::Ip);
+    auto* socket_address = proto_address.mutable_socket_address();
+    socket_address->set_address(address.ip()->addressAsString());
+    socket_address->set_port_value(address.ip()->port());
+  }
+}
+
+void ExtAuthzCheckRequestGenerator::setAttrContextPeer(
+    envoy::service::auth::v2::AttributeContext_Peer& peer, const Network::Connection& connection,
+    const std::string& service, const bool local) {
+
+  // Set the address
+  auto addr = peer.mutable_address();
+  if (!local) {
+    addressToProtobufAddress(*addr, *connection.remoteAddress());
+  } else {
+    addressToProtobufAddress(*addr, *connection.localAddress());
+  }
+
+  // Set the principal
+  // Preferably the SAN from the peer's cert or
+  // Subject from the peer's cert.
+  std::string principal;
+  Ssl::Connection* ssl = const_cast<Ssl::Connection*>(connection.ssl());
+  if (ssl != nullptr) {
+    if (!local) {
+      principal = ssl->uriSanPeerCertificate();
+
+      if (principal.empty()) {
+        principal = ssl->subjectPeerCertificate();
+      }
+    } else {
+      principal = ssl->uriSanLocalCertificate();
+
+      if (principal.empty()) {
+        principal = ssl->subjectLocalCertificate();
+      }
+    }
+  }
+  peer.set_principal(principal);
+
+  if (!service.empty()) {
+    peer.set_service(service);
+  }
+}
+
+const std::string ExtAuthzCheckRequestGenerator::getProtocolStr(const Envoy::Http::Protocol& p) {
```

Comment:
`const` qualifier here is not needed (it doesn't strengthen the type). Also, prefer `getProtocolString` as a nit, we try and avoid abbreviations where things don't get massively verbose.